leap/multistep/__init__.py

```diff
@@ -416,4 +417,341 @@ def rk_bootstrap(self, cb):
+                dt_factor = self.dt
+                t_end = 1
+
+            # Implicit setup - rhs_next_var is an unknown, needs implicit solve.
+            equations = []
+            unknowns = set()
+            knowns = set()
+
+            unknowns.add(rhs_next_var)
+
+            # Update history
+            history = self.history + [rhs_next_var]
+
+            # Set up the actual Adams-Moulton step.
+            ab_sum = emit_adams_integration(
+                            cb_primary, name_gen,
+                            self.function_family,
+                            time_hist, history,
+                            0, t_end)
+
+            state_est = self.state + dt_factor * ab_sum
+            if self.state_filter is not None:
+                state_est = self.state_filter(state_est)
+
+            # Build the implicit solve expression.
+            from dagrt.expression import collapse_constants
+            from pymbolic.mapper.distributor import DistributeMapper as DistMap
+            solve_expression = collapse_constants(
+                    rhs_next_var - self.eval_rhs(self.t + self.dt,
+                                                 DistMap()(state_est)),
+                    list(unknowns) + [self.state],
+                    cb_primary.assign, cb_primary.fresh_var)
+            equations.append(solve_expression)
+
+            # {{{ emit solve if possible
+
+            if unknowns and len(unknowns) == len(equations):
+                # got a square system, let's solve
+                assignees = [unk.name for unk in unknowns]
+
+                from pymbolic import substitute
+                subst_dict = dict(
+                        (rhs_var.name, rhs_var_to_unknown[rhs_var])
+                        for rhs_var in unknowns)
+
+                cb_primary.assign_implicit(
+                        assignees=assignees,
+                        solve_components=[
+                            rhs_var_to_unknown[unk].name
+                            for unk in unknowns],
+                        expressions=[
+                            substitute(eq, subst_dict)
+                            for eq in equations],
+
+                        # TODO: Could supply a starting guess
+                        other_params={
+                            "guess": self.state},
+                        solver_id="solve")
+
+                del equations[:]
+                knowns.update(unknowns)
+                unknowns.clear()
+
+            # }}}
+
+            # Update the state now that we've solved.
+            cb_primary(self.state, state_est)
+
+            # Rotate history and time history.
+            for i in range(self.hist_length - 1):
+                cb_primary(self.history[i], history[i + 1])
+
+                if not self.static_dt:
+                    cb_primary(self.time_history[i], time_history_data[i + 1])
```

Comment:
I imagine this time history management code occurs similarly somewhere else. Could you factor this out into a utility?